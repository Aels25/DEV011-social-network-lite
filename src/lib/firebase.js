// firebase.js
import { initializeApp } from 'firebase/app';
import {
  getAuth,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signInWithPopup,
  GoogleAuthProvider,
} from 'firebase/auth';

// Configuraci√≥n de Firebase
const firebaseConfig = {
  apiKey: 'AIzaSyAyFaL_gwYA_0KzH4m4T9h6ftaHo0bD4_E',
  authDomain: 'dev011-social-network-lite.firebaseapp.com',
  projectId: 'dev011-social-network-lite',
  storageBucket: 'dev011-social-network-lite.appspot.com',
  messagingSenderId: '316146436395',
  appId: '1:316146436395:web:19537a224689c6b9509d6e',
  measurementId: 'G-20XZP5RSX1',
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
const googleProvider = new GoogleAuthProvider();

export {
  auth, googleProvider, createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup,
};
