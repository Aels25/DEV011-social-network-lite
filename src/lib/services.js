// Esta función debería borrar todos los datos del almacenamiento local
export const init = () => {
  localStorage.clear();
}

// Esta función debería devolver true si el usuario existe
// Debería devolver false si el usuario no existe
export const login = (email, password) => {
  const usersStr = localStorage.getItem("users");
  if(usersStr){
    const users = JSON.parse(usersStr);
    const user = users.find(user => user.email === email && user.password === password);
    if(user){
      localStorage.setItem("user", JSON.stringify(user));
      return true;
    }
  }
  return false;
}

// Esta función debería devolver al usuario que ha iniciado sesión
// Debería devolver null si no hay usuario iniciado sesión
export const getLoggedInUser = () => {
  const userStr = localStorage.getItem("user");
  return userStr ? JSON.parse(userStr) : null;
}

// Esta función debería cerrar la sesión del usuario
export const logout = () => {
  localStorage.removeItem("user");
}

// Esta función debería registrar a un nuevo usuario
// Debería devolver true si el usuario se registró con éxito
// Debería lanzar un error si el usuario ya existe
export const register = (email, password) => {
  // Verificar la expresión regular del correo electrónico
  const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  if(!emailRegex.test(email)){
    throw new Error("Correo electrónico no válido");
  }

  // Verificar la longitud de la contraseña
  if(password.length < 6){
    throw new Error("La contraseña debe tener al menos 6 caracteres");
  }

  // Verificar si el usuario ya existe
  let users = [];
  const usersStr = localStorage.getItem("users");
  if(usersStr){
    users = JSON.parse(usersStr);
  }
  const user = users.find(user => user.email === email);
  if(user){
    throw new Error("El usuario ya existe");
  }
  else{
    users.push({email, password});
    localStorage.setItem("users", JSON.stringify(users));
    return true;
  }
}

// Esta función debería devolver un array de posts
// Cada post debería tener la siguiente estructura:
// {
//   id: string,
//   content: string,
//   email: string
// }
export const getPosts = () => {
  const postsStr = localStorage.getItem("posts");
  return postsStr ? JSON.parse(postsStr) : [];
}

// Esta función debería crear un nuevo post y devolver su id
// El post debería tener la siguiente estructura:
// {
//   id: string,
//   content: string,
//   email: string
// }
export const createPost = (content, email) => {
  // Verificar la longitud del contenido
  if(content.length < 1){
    throw new Error("El contenido debe tener al menos 1 carácter");
  }

  // Verificar la expresión regular del correo electrónico
  const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  if(!emailRegex.test(email)){
    throw new Error("Correo electrónico no válido");
  }

  const id = Math.random().toString(36).substr(2, 9);
  let posts = [];
  const postsStr = localStorage.getItem("posts");
  if(postsStr){
    posts = JSON.parse(postsStr);
  }
  posts.push({id, content, email});
  localStorage.setItem("posts", JSON.stringify(posts));
  return id;
}

// Esta función debería editar el contenido de un post
// Debería lanzar un error si el post no existe
export const editPost = (idPost, content) => {
  // Verificar la longitud del contenido
  if(content.length < 1){
    throw new Error("El contenido debe tener al menos 1 carácter");
  }

  const postsStr = localStorage.getItem("posts");
  if(postsStr){
    const posts = JSON.parse(postsStr);
    const post = posts.find(post => post.id === idPost);
    if(post){
      post.content = content;
      localStorage.setItem("posts", JSON.stringify(posts));
    }
    else{
      throw new Error("El post no existe");
    }
  }
  else{
    throw new Error("El post no existe");
  }
}

// Esta función debería eliminar un post
// Debería lanzar un error si el post no existe
export const deletePost = (idPost) => {
  const postsStr = localStorage.getItem("posts");
  if(postsStr){
    const posts = JSON.parse(postsStr);
    const post = posts.find(post => post.id === idPost);
    if(post){
      const index = posts.indexOf(post);
      posts.splice(index, 1);
      localStorage.setItem("posts", JSON.stringify(posts));
    }
    else{
      throw new Error("El post no existe");
    }
  }
  else{
    throw new Error("El post no existe");
  }
}
export const myFunction = () => {
  // lógica de myFunction
};