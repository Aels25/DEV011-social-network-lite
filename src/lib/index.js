
// Configuración de la autenticación con Google
import { auth, googleProvider, signInWithPopup } from "./firebase";


document.getElementById("google-auth-btn").addEventListener("click", () => {
  signInWithPopup(auth, googleProvider)
    .then((result) => {
      // El usuario ha iniciado sesión correctamente
      const user = result.user;
      console.log("Usuario autenticado con Google:", user);
    })
    .catch((error) => {
      // Manejar errores aquí
      const errorCode = error.code;
      const errorMessage = error.message;
      console.error("Error al autenticar con Google:", errorCode, errorMessage);
    });
});
