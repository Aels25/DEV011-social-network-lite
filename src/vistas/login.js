// file login.js
import { auth, signInWithEmailAndPassword } from '../lib/firebase'; // Importa las funciones necesarias de Firebase

function login(navigateTo) {
  const section = document.createElement('section');
  const title = document.createElement('h2');
  const buttonReturn = document.createElement('button');
  const form = document.createElement('form');
  const inputEmail = document.createElement('input');
  const inputPass = document.createElement('input');
  const buttonLogin = document.createElement('button');
  const errorContainer = document.createElement('div'); // Nuevo elemento para mostrar errores

  form.id = 'login-form';
  inputEmail.id = 'login-inputEmail';
  inputPass.id = 'login-inputPass';
  buttonReturn.id = 'btnLoginReturn';
  buttonLogin.id = 'btnLogin';

  inputEmail.placeholder = 'Escribe tu email';
  inputPass.placeholder = 'Contraseña';

  title.textContent = 'Conéctate a WANDERWEB';
  buttonLogin.textContent = 'Iniciar Sesión';

  buttonReturn.textContent = 'Volver a la página de inicio';
  buttonReturn.addEventListener('click', () => {
    navigateTo('/');
  });

  // Evento para el formulario de inicio de sesión
  form.addEventListener('submit', async (event) => {
    event.preventDefault();

    const email = inputEmail.value;
    const password = inputPass.value;

    try {
      // Utiliza la función de Firebase para iniciar sesión con email y contraseña
      await signInWithEmailAndPassword(auth, email, password);

      // Redirige a la vista del muro después de iniciar sesión
      navigateTo('/muro');
    } catch (error) {
      // Muestra el mensaje de error en el nuevo contenedor de errores
      errorContainer.textContent = `Error al iniciar sesión: ${error.message}`;
    }
  });

  form.append(inputEmail, inputPass, buttonLogin);
  section.append(title, form, buttonReturn, errorContainer);

  return section;
}

export default login;
