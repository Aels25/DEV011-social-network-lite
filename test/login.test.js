/**
 * @jest-environment jsdom
 */

import login from '../src/vistas/login.js';
import { signInWithEmailAndPassword } from '../src/lib/firebase';

jest.mock('../src/lib/firebase', () => ({
  auth: {},
  signInWithEmailAndPassword: jest.fn(),
}));

describe('login', () => {
  test('snapshot of login', () => {
    const DOM = document.createElement('div');
    DOM.append(login(() => {}));
    expect(DOM).toMatchSnapshot();
  });

  test('is a function', () => {
    expect(typeof login).toBe('function');
  });

  test('have a button Return', () => {
    const DOM = document.createElement('div');
    DOM.append(login(() => {}));
    const haveAButton = DOM.querySelector('#btnLoginReturn');
    expect(haveAButton).not.toBe(null);
  });

  test('have a button Login', () => {
    const DOM = document.createElement('div');
    DOM.append(login(() => {}));
    const haveAButton = DOM.querySelector('#btnLogin');
    expect(haveAButton).not.toBe(null);
  });

  test('button Return calls navigateTo on click', () => {
    const navigateTo = jest.fn();
    const DOM = document.createElement('div');
    DOM.append(login(navigateTo));
    const buttonReturn = DOM.querySelector('#btnLoginReturn');
    buttonReturn.click();
    expect(navigateTo).toHaveBeenCalledTimes(1);
    expect(navigateTo).toHaveBeenCalledWith('/');
  });

  test('form submission calls signInWithEmailAndPassword', async () => {
    const navigateTo = jest.fn();
    const DOM = document.createElement('div');
    DOM.append(login(navigateTo));
    const form = DOM.querySelector('#login-form');
    const email = DOM.querySelector('#login-inputEmail');
    const password = DOM.querySelector('#login-inputPass');

    email.value = 'test@example.com';
    password.value = 'password123';

    await form.dispatchEvent(new Event('submit'));

    expect(signInWithEmailAndPassword).toHaveBeenCalledTimes(1);
    expect(signInWithEmailAndPassword).toHaveBeenCalledWith({}, 'test@example.com', 'password123');
  });
});
