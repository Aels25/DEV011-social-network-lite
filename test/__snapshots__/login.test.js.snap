// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login snapshot of login 1`] = `
<div>
  <section>
    <h2>
      Conéctate a WANDERWEB
    </h2>
    <form
      id="login-form"
    >
      <input
        id="login-inputEmail"
        placeholder="Escribe tu email"
      />
      <input
        id="login-inputPass"
        placeholder="Contraseña"
      />
      <button
        id="btnLogin"
      >
        Iniciar Sesión
      </button>
    </form>
    <button
      id="btnLoginReturn"
    >
      Volver a la página de inicio
    </button>
    <div />
  </section>
</div>
`;
